README.md
[![Downloads](https://img.shields.io/github/downloads/diasurgical/devilutionX/total.svg)](https://github.com/diasurgical/devilutionX/releases)
[![github stars](https://img.shields.io/github/stars/diasurgical/devilutionX.svg)](https://github.com/diasurgical/devilutionX/stargazers)
[![codecov](https://codecov.io/gh/diasurgical/devilutionX/branch/master/graph/badge.svg)](https://codecov.io/gh/diasurgical/devilutionX)

Nightly builds | Platform
---:| ---
[![CircleCI](https://circleci.com/gh/diasurgical/devilutionX.svg?style=svg)](https://circleci.com/gh/diasurgical/devilutionX) | Linux 32bit & 64bit, Windows 32bit, Nintendo Switch, SDL1
[![Build Status](https://travis-ci.org/diasurgical/devilutionX.svg?branch=master)](https://travis-ci.org/diasurgical/devilutionX) | macOS 64bit
[![Build status](https://ci.appveyor.com/api/projects/status/1a0jus2372qvksht?svg=true)](https://ci.appveyor.com/project/AJenbo/devilutionx) | Windows 64bit (MSVC)

![Discord Channel](https://avatars3.githubusercontent.com/u/1965106?s=16&v=4) [Discord Chat Channel](https://discord.gg/aQBQdDe)

# How To Play:
 - Download [the latest DevilutionX release](https://github.com/diasurgical/devilutionX/releases), or build from source
 - Copy diabdat.mpq from your CD or GoG installation (or [extract it from the GoG installer](https://github.com/diasurgical/devilutionX/wiki/Extracting-the-DIABDAT.MPQ-from-the-GoG-installer)) to the DevilutionX install folder or data folder; make sure it is all lowercase. The DeviliutionX install folder is the one that contains the DevilutionX executable. The data folder path may differ depending on OS version and security settings, but will normally be as follows:
    - macOS `~/Library/Application Support/diasurgical/devilution`
    - Linux `~/.local/share/diasurgical/devilution/`
    - Windows `C:\Users\[username]\AppData\Roaming\diasurgical\devilution`
 - Install [SDL2](https://www.libsdl.org/download-2.0.php), [SDL2_mixer](https://www.libsdl.org/projects/SDL_mixer/) and [SDL2_ttf](https://www.libsdl.org/projects/SDL_ttf/) (included in macOS and Windows releases):
    - Ubuntu/Debian/Rasbian `sudo apt-get install libsdl2-ttf-2.0-0 libsdl2-mixer-2.0-0`
 - Run `./devilutionx`

# Building from Source
<details><summary>Linux</summary>

### Installing dependencies on Debian and Ubuntu
```
sudo apt-get install cmake g++ libsdl2-mixer-dev libsdl2-ttf-dev libsodium-dev
```
### Installing dependencies on Fedora
```
sudo dnf install cmake glibc-devel SDL2-devel SDL2_ttf-devel SDL2_mixer-devel libsodium-devel libasan libubsan
```
### Compiling
```
cd build
cmake ..
make -j$(nproc)
```
</details>

<details><summary>macOS</summary>

Make sure you have [Homebrew](https://brew.sh/) installed, then run:

```
brew bundle
cd build
cmake ..
cmake --build . -j $(sysctl -n hw.physicalcpu)
```
</details>
<details><summary>FreeBSD</summary>

### Installing dependencies
```
pkg install cmake sdl2_mixer sdl2_ttf libsodium
```
### Compiling
```
cd build
cmake ..
cmake --build . -j $(sysctl -n hw.ncpu)
```
</details>

<details><summary>OpenBSD</summary>

### Installing dependencies
```
pkg_add cmake sdl2-mixer sdl2-ttf libsodium gmake
```
### Compiling
```
cd build
cmake -DCMAKE_MAKE_PROGRAM=gmake ..
cmake --build . -j $(sysctl -n hw.ncpuonline)
```
</details>

<details><summary>Windows via MinGW</summary>

### Installing dependencies on WSL, Debian and Ubuntu

Download and place the 32bit MinGW Development Libraries of [SDL2](https://www.libsdl.org/download-2.0.php), [SDL2_mixer](https://www.libsdl.org/projects/SDL_mixer/), [SDL2_ttf](https://www.libsdl.org/projects/SDL_ttf/) and [Libsodium](https://github.com/jedisct1/libsodium/releases) in `/usr/i686-w64-mingw32`.

```
sudo apt-get install cmake gcc-mingw-w64-i686 g++-mingw-w64-i686
```
### Compiling
```
cd build
cmake -DCMAKE_TOOLCHAIN_FILE=../CMake/mingwcc.cmake ..
make -j$(nproc)
```
</details>
<details><summary>Windows via Visual Studio</summary>

### Installing dependencies
Make sure to install the `C++ CMake tools for Windows` component for Visual Studio.

* **Using vcpkg (recommended)**
1. Install vcpkg following the instructions from https://github.com/microsoft/vcpkg#quick-start.

   Don't forget to perform _user-wide integration_ step for additional convenience.
2. Install required dependencies by executing the following command (via cmd or powershell):

   For the 64-bit version of the dependencies please run this command:

   ```
   vcpkg install sdl2:x64-windows sdl2-mixer:x64-windows sdl2-ttf:x64-windows libsodium:x64-windows
   ```

   For the 32-bit version of the dependencies please run this command:

   ```
   vcpkg install sdl2:x86-windows sdl2-mixer:x86-windows sdl2-ttf:x86-windows libsodium:x86-windows
   ```

* **Manually**
1. Download and place the MSVC Development Libraries of [SDL2](https://www.libsdl.org/download-2.0.php), [SDL2_mixer](https://www.libsdl.org/projects/SDL_mixer/), [SDL2_ttf](https://www.libsdl.org/projects/SDL_ttf/) and [Libsodium](https://github.com/jedisct1/libsodium/releases) in `%USERPROFILE%\AppData\Local\Microsoft\WindowsApps\`.
2. If dependencies are not found or you wish to place them in other location - configure required path variables in _"Manage Configurations..."_ dialog inside Visual Studio or in _cmake-gui_.

### Compiling

* **Through Open->CMake in Visual Studio**
1. Go to `File -> Open -> CMake`, select `CMakeLists.txt` from the project root.
2. Select `Build devilution.exe` from the `Build` menu.

* **Through cmake-gui**

1. Input the path to devilutionx source directory at `Where is the source code:` field.
2. Input the path where the binaries would be placed at `Where to build the binaries:` field. If you want to place them inside source directory it's preferable to do so inside directory called `build` to avoid the binaries being added to the source tree.
3. It's recommended to input `Win32` in `Optional Platform for Generator`, otherwise it will default to x64 build.
4. In case you're using `vcpkg` select `Specify toolchain file for cross-compiling` and select the file `scripts/buildsystems/vcpkg.cmake` from `vcpkg` directory otherwise just go with `Use default native compilers`.
5. In case you need to select any paths to dependencies manually do this right in cmake-gui window.
6. Press `Generate` and open produced `.sln` file using Visual Studio.
7. Use build/debug etc. commands inside Visual Studio Solution like with any normal Visual Studio project.
</details>

<details><summary>Nintendo Switch</summary>
Run:

```
Packaging/switch/build.sh
```

This will install the [Switch devkit](https://switchbrew.org/wiki/Setting_up_Development_Environment) and build a DevilutionX Switch package. If you already have the devkit installed, or are on a non-Debian system, pass the the devkit path to the script like this:

```
DEVKITPRO=<path to devkit> Packaging/switch/build.sh
```

The nro-file will be generated in the build folder. Test with an emulator (RyuJinx) or real hardware.
</details>

<details><summary>Haiku</summary>

### Installing dependencies on 32 bit Haiku
```
pkgman install cmake_x86 devel:libsdl2_x86 devel:libsdl2_mixer_x86 devel:libsdl2_ttf_x86 devel:libsodium_x86
```
### Installing dependencies on 64 bit Haiku
```
pkgman install cmake devel:libsdl2 devel:libsdl2_mixer devel:libsdl2_ttf devel:libsodium
```
### Compiling on 32 bit Haiku
```
cd build
setarch x86 #Switch to secondary compiler toolchain (GCC8+)
cmake ..
cmake --build . -j $(nproc)
```
### Compiling on 64 bit Haiku
No setarch required, as there is no secondary toolchain on x86_64, and the primary is GCC8+
```
cd build
cmake ..
cmake --build . -j $(nproc)
```
</details>

<details><summary>OpenDingux / RetroFW</summary>

DevilutionX uses buildroot to build packages for OpenDingux and RetroFW.

The build script does the following:

1. Downloads and configures the buildroot if necessary.
2. Builds the executable (using CMake).
3. Packages the executable and all related resources into an `.ipk` or `.opk` package.

The buildroot uses ~2.5 GiB of disk space and can take 20 minutes to build.

For OpenDingux builds `mksquashfs` needs to be installed.

To build, run the following command

~~~ bash
Packaging/OpenDingux/build.sh <platform>
~~~

Replace `<platform>` with one of: `retrofw`, `rg350`, or `gkd350h`.

This prepares and uses the buildroot at `$HOME/buildroot-$PLATFORM-devilutionx`.

</details>

<details><summary>Clockwork PI GameShell</summary>

You can either call
~~~ bash
Packaging/cpi-gamesh/build.sh
~~~
to install dependencies and build the code.

Or you create a new directory under `/home/cpi/apps/Menu` and copy [the file](Packaging/cpi-gamesh/__init__.py) there. After restarting the UI, you can download and compile the game directly from the device itself. See [the readme](Packaging/cpi-gamesh/readme.md) for more details.
</details>

<details><summary>Amiga via Docker</summary>

### Build the container from the repo root

~~~ bash
docker build -f Packaging/amiga/Dockerfile -t devilutionx-amiga .
~~~

### Build DevilutionX Amiga binary

~~~ bash
docker run --rm -v "${PWD}:/work" devilutionx-amiga
sudo chown "${USER}:" build-amiga/*
~~~

The command above builds DevilutionX in release mode.
For other build options, you can run the container interactively:

~~~ bash
docker run -ti --rm -v "${PWD}:/work" devilutionx-amiga bash
~~~

See the `CMD` in `Packaging/amiga/Dockerfile` for reference.

### Copy the necessary files

Outside of the Docker container, from the DevilutionX directory, run:

~~~ bash
cp Packaging/amiga/devilutionx.info Packaging/amiga/LiberationSerif-Bold.ttf build-amiga/
sudo chown "${USER}:" build-amiga/*
~~~

To actually start DevilutionX, increase the stack size to 50KiB in Amiga.
You can do this by selecting the DevilutionX icon, then hold right mouse button and
select Icons -> Information in the top menu.
</details>

<details><summary><b>CMake build options</b></summary>

### General
- `-DBINARY_RELEASE=ON` changed build type to release and optimize for distribution.
- `-DNONET=ON` disable network support, this also removes the need for the ASIO and Sodium.
- `-DUSE_SDL1=ON` build for SDL v1 instead of v2, not all features are supported under SDL v1, notably upscaling.
- `-DSPAWN=ON` build the shareware version, using spawn.mpq from the original shareware; which can still be [downloaded](http://ftp.blizzard.com/pub/demos/diablosw.exe) for free.
- `-DCMAKE_TOOLCHAIN_FILE=../CMake/32bit.cmake` generate 32bit builds on 64bit platforms (remember to use the `linux32` command if on Linux).
- `-DCROSS_PREFIX=/path/to/prefix` set the path to the `i686-w64-mingw32` directory.

### Debug builds
- `-DDEBUG=OFF` disable debug mode of the Diablo engine.
- `-DASAN=OFF` disable address sanitizer.
- `-DUBSAN=OFF` disable undefined behavior sanitizer.

</details>

# Multiplayer
 - TCP/IP only requires the host to expose port 6112
 - UDP/IP requires that all players expose port 6112

All games are encrypted and password protected.

# Controller support

DevilutionX supports gamepad controls.

Default controller mappings (A/B/X/Y as in Nintendo layout, so the rightmost button is attack):

- Left analog or D-Pad: move hero
- A: attack nearby enemies, talk to townspeople and merchants, pickup/place items in the inventory, OK while in main menu
- B: select spell, back while in menus
- X: pickup items, open nearby chests and doors, use item in the inventory
- Y: cast spell, delete character while in main menu
- L1: use health item from belt
- R1: use mana potion from belt
- L2 or Start + Left: character sheet
- R2 or Start + Right: inventory
- Left analog click or Start + Down: toggle automap
- Start + Up or Start + Select: game menu
- Start + Left: character info
- Start + Right: inventory
- Select + A/B/X/Y: Spell hotkeys
- Right analog: move automap or simulate mouse
- Right analog click or Select + L1: left mouse click
- Select + Right analog click or Select + R1: right mouse click
- Select + L2 or Start + Y: quest log
- Select + R2 or Start + B: spell book

For now, they can be re-mapped by changing `SourceX/controls` or by setting the `SDL_GAMECONTROLLERCONFIG` environment
variable (see
[SDL_GameControllerDB](https://github.com/gabomdq/SDL_GameControllerDB)).

# Contributing
[Guidelines](docs/CONTRIBUTING.md)

# Mods

[List of known mods based on DevilutionX](docs/mods.md)

# F.A.Q.
> Wow, does this mean I can download and play Diablo for free now?

No, you'll need access to the data from the original game. If you don't have an original CD then you can [buy Diablo from GoG.com](https://www.gog.com/game/diablo). Alternately you can use `spawn.mpq` from the [http://ftp.blizzard.com/pub/demos/diablosw.exe](shareware) version and compile the with the SPAWN flag defined.
> What game changes does DevilutionX provide

DevilutionX's main focus is to make the game work on multiple platforms. An additional goal is to make the engine mod friendly. As such, there are no changes to gameplay, but we will be making some enhancments to the engine itself. For example, the engine now has upscaling, unlocked fps, controller support, and multiplayer via TCP.
> Is 1080p supported?

Currently the game simply scales the original 640x480 to best fit, but we are working on widescreen support.
> What about Hellfire?

Hellfire is being worked on and is mostly done, though not fully playable at the moment.
> Does it work with Battle.net?

Battle.net is a service provided by Blizzard. We are not associated with them, so we have not worked on intergrating with their service.
</details>

# Credits
- The original Devilution project [Devilution](https://github.com/diasurgical/devilution#credits)
- [Everyone](https://github.com/diasurgical/devilutionX/graphs/contributors) who worked on Devilution/DevilutionX
- And a thanks to all who support the project, report bugs and help spread the word <3

# Changelog
[From the beginning until release](docs/CHANGELOG.md)

# Legal
DevilutionX is released to the Public Domain. The documentation and functionality provided by Devilution may only be utilized with assets provided by ownership of Diablo.

Diablo® - Copyright © 1996 Blizzard Entertainment, Inc. All rights reserved. Diablo and Blizzard Entertainment are trademarks or registered trademarks of Blizzard Entertainment, Inc. in the U.S. and/or other countries.

Devilution and any of its maintainers are in no way associated with or endorsed by Blizzard Entertainment®.


docs/CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## DevilutionX 1.0.0
### Features
- FPS no longer capped at 20hz; smoother mouse and transitions
- Full-featured gamepad support
- All movies can now be skipped via ESC
- Implement `--help`, `--version` and `--data-dir`
- Implement FPS counter (`-f`)
- Force windowed mode with `-x`
- Skip startup movies with `-n`
- Implement scroll wheel navigation
- Implement touch support
- Load assets from the same folder as the mpq
- Faster loading

### Platforms
- Added Nintendo Switch support
- Added Retro Gamer Handheld (OpenDingux/Retrofw) support
- Added OpenBSD support
- Windows version is now build as GUI an app and has an icon

### Bugfixes
- Fixed random missing sounds
- Fixed NPCs not talking about the right quest
- Error messages implemented in GUI
- Fixed having to restart the game between network sessions
- Fixed game not working on some Radeon GPUs
- Fixed panel missing after minimizing game if upscaling is disabled
- Fixed clicks in the left letterbox being incorrectly handled
- Fixed end movie not looping

### Original Diablo bugs
- Fixed a large number of rare stability issues from the original game
- Correctly take open panels into account when casting spells
- Fix sound some times being muted after Diablo dies

## DevilutionX 0.5.0
### Features
- Sound is now accurate to the original
- All in-game issues fixed
- Delete hero, inline dialogs and scrollbars are now implemented
- Screenshots now have different names
- Multiple simultaneous dialogs fixed
- All builds are now 64bit (except for Windows and Raspberry Pi)
- Memory leaks and crashes fixed
- All keys are now mapped
- UI text now has correct shadows
- Much lower CPU usage
- diabdat.mpq can now be loaded with read-only access

### Known issues
- Error dialogs not implemented in main UI
- The game must restart after hosting multiplayer

## [0.10.0](https://github.com/diasurgical/devilution/compare/0.9.6...0.10.0)
### June 22, 2019
- [All functions are now binary identical](https://github.com/diasurgical/devilution/milestone/3) to the 1.09b version
- Fix buying from Wirt
- Replace many magic numbers with constants
- Fix a handful of minor issues
- Add toggle fullscreen with alt+enter in debug builds

### June 21, 2019
- All functions are now [binary identical](https://github.com/diasurgical/devilution/milestone/3) to Diablo 1.09b

### June 1, 2019
- MVG posts [a video about the project](https://www.youtube.com/watch?v=5tADL_fmsHQ) and releases a Nintendo Switch port

## DevilutionX 0.4.0
### May 20, 2019
- Fixed flickering mouse in caves
- 32bit ARM build (Raspberry Pi)
- 32bit FreeBSD build
- 32bit Haiku support (see HaikuDepot)
- Included font for rendering credits and progress screen
- Upscaling quality can be adjusted or fully disabled in diablo.ini
- Windowed mode can be set in diablo.ini
- Mouse capture can be set in diablo.ini
- Direct file access implemented (mods won't need to pack their files in an MPQ)
- Music and speech memory leaks fixed

## [0.9.6](https://github.com/diasurgical/devilution/compare/0.9.0...0.9.6)
### May 19, 2019
- [96% of functions are now binary identical](https://github.com/diasurgical/devilution/milestone/3) to the 1.09b version
- Fix several item corruption issues introduced in 0.9.0

## [0.9.0](https://github.com/diasurgical/devilution/compare/0.8.0...0.9.0)
### May 2, 2019
- [90% of functions are now binary identical](https://github.com/diasurgical/devilution/milestone/3) to the 1.09b version
- Mute buttons now work correctly

### April 15, 2019
- Code is once again compiled as C++ as some parts appear to require despite the indications in Rich header

## [0.8.0](https://github.com/diasurgical/devilution/compare/0.7.0...0.8.0)
### April 12, 2019
- [80% of functions are now binary identical](https://github.com/diasurgical/devilution/milestone/3) to the 1.09b version
- Fixes a few minor issues with generated items

## [0.7.0](https://github.com/diasurgical/devilution/compare/0.6.0...0.7.0)
### April 9, 2019
- [70% of functions are now binary identical](https://github.com/diasurgical/devilution/milestone/3) to the 1.09b version

### April 9, 2019
- The last of the compiler flags are figured out

### March 22, 2019
- Devilution appears on [Phoronix](https://www.phoronix.com/scan.php?page=news_item&px=DeviluitionX-Open-Diablo)

## DevilutionX 0.3.0
### March 20, 2019
- Fix dialog volume
- 32bit macOS build
- Fix crash in town during multiplayer
- Screenshot implemented

## [0.6.0](https://github.com/diasurgical/devilution/compare/v0.5.0...0.6.0)
### March 19, 2019
- [60% of functions are now binary identical](https://github.com/diasurgical/devilution/milestone/3) to the 1.09b version
- Added a guide for people wanting to join in [Cleaning the code](https://github.com/diasurgical/devilution/wiki/Cleaning-Code)
- File size is now only 968 bytes (0.13%) larger than the original Diablo 1.09b.

## DevilutionX 0.2.0
### March 17, 2019
- Fully implemented audio (all issues from 0.1.0 fixed)
- Fully implemented multiplayer
- Windows 32bit build
- Gameplay is fully featured and plays like the original
- Memory leaks fixed
- Most known crashes have been fixed

### March 7, 2019
- [GOG re-release Diablo](https://www.gog.com/news/release_diablo)

## DevilutionX 0.1.0
### February 27, 2019
- Linux 32bit build
- Basic audio
- Video playback
- Basic menus
- Graphics
- Gameplay
- Basic multiplayer over TCP/UDP
- Persistent settings
- Encrypted network connection with password protection

## [0.5.0](https://github.com/diasurgical/devilution/compare/0.4...v0.5.0)
### January 14, 2019
- [50% of functions are now binary identical](https://github.com/diasurgical/devilution/milestone/3) to the 1.09b version

sha.asm
; Listing generated by Microsoft (R) Optimizing Compiler Version 12.00.9044.0 

	TITLE	Source/sha.cpp
	.386P
include listing.inc
if @Version gt 510
.model FLAT
else
_TEXT	SEGMENT PARA USE32 PUBLIC 'CODE'
_TEXT	ENDS
_DATA	SEGMENT DWORD USE32 PUBLIC 'DATA'
_DATA	ENDS
CONST	SEGMENT DWORD USE32 PUBLIC 'CONST'
CONST	ENDS
_BSS	SEGMENT DWORD USE32 PUBLIC 'BSS'
_BSS	ENDS
$$SYMBOLS	SEGMENT BYTE USE32 'DEBSYM'
$$SYMBOLS	ENDS
$$TYPES	SEGMENT BYTE USE32 'DEBTYP'
$$TYPES	ENDS
_TLS	SEGMENT DWORD USE32 PUBLIC 'TLS'
_TLS	ENDS
CRT$XCU	SEGMENT DWORD USE32 PUBLIC ''
CRT$XCU	ENDS
;	COMDAT _SHA1Clear
_TEXT	SEGMENT PARA USE32 PUBLIC 'CODE'
_TEXT	ENDS
;	COMDAT @SHA1Result@8
_TEXT	SEGMENT PARA USE32 PUBLIC 'CODE'
_TEXT	ENDS
;	COMDAT @SHA1Calculate@12
_TEXT	SEGMENT PARA USE32 PUBLIC 'CODE'
_TEXT	ENDS
;	COMDAT @SHA1Input@12
_TEXT	SEGMENT PARA USE32 PUBLIC 'CODE'
_TEXT	ENDS
;	COMDAT @SHA1ProcessMessageBlock@4
_TEXT	SEGMENT PARA USE32 PUBLIC 'CODE'
_TEXT	ENDS
;	COMDAT @SHA1Reset@4
_TEXT	SEGMENT PARA USE32 PUBLIC 'CODE'
_TEXT	ENDS
;	COMDAT @SHA1Init@4
_TEXT	SEGMENT PARA USE32 PUBLIC 'CODE'
_TEXT	ENDS
FLAT	GROUP _DATA, CONST, _BSS, CRT$XCU
	ASSUME	CS: FLAT, DS: FLAT, SS: FLAT
endif

INCLUDELIB LIBCMT
INCLUDELIB OLDNAMES

_DATA	SEGMENT
COMM	_sgSHA1:BYTE:0114H
_DATA	ENDS
PUBLIC	_SHA1Clear
EXTRN	_memset:NEAR
; Function compile flags: /Ogsy
;	COMDAT _SHA1Clear
_TEXT	SEGMENT
_SHA1Clear PROC NEAR					; COMDAT
; File Source/sha.cpp
; Line 9
	push	276					; 00000114H
	push	0
	push	OFFSET FLAT:_sgSHA1
	call	_memset
	add	esp, 12					; 0000000cH
; Line 10
	ret	0
_SHA1Clear ENDP
_TEXT	ENDS
PUBLIC	@SHA1Result@8
; Function compile flags: /Ogsy
;	COMDAT @SHA1Result@8
_TEXT	SEGMENT
@SHA1Result@8 PROC NEAR					; COMDAT
; _n$ = ecx
; _Message_Digest$ = edx
; Line 13
	mov	eax, edx
; Line 18
	test	eax, eax
	je	SHORT $L84995
; Line 19
	imul	ecx, 92					; 0000005cH
	push	5
	add	ecx, OFFSET FLAT:_sgSHA1
	pop	edx
	push	esi
$L84993:
; Line 20
	mov	esi, DWORD PTR [ecx]
	add	ecx, 4
	mov	DWORD PTR [eax], esi
; Line 21
	add	eax, 4
	dec	edx
	jne	SHORT $L84993
	pop	esi
$L84995:
; Line 24
	ret	0
@SHA1Result@8 ENDP
_TEXT	ENDS
PUBLIC	@SHA1Calculate@12
PUBLIC	@SHA1Input@12
; Function compile flags: /Ogsy
;	COMDAT @SHA1Calculate@12
_TEXT	SEGMENT
_Message_Digest$ = 8
@SHA1Calculate@12 PROC NEAR				; COMDAT
; _n$ = ecx
; _data$ = edx
; Line 27
	push	esi
	mov	esi, ecx
; Line 28
	imul	ecx, 92					; 0000005cH
	push	64					; 00000040H
	add	ecx, OFFSET FLAT:_sgSHA1
	call	@SHA1Input@12
; Line 29
	mov	edx, DWORD PTR _Message_Digest$[esp]
	test	edx, edx
	je	SHORT $L85003
; Line 30
	mov	ecx, esi
	call	@SHA1Result@8
$L85003:
	pop	esi
; Line 31
	ret	4
@SHA1Calculate@12 ENDP
_TEXT	ENDS
PUBLIC	@SHA1ProcessMessageBlock@4
EXTRN	_memcpy:NEAR
; Function compile flags: /Ogsy
;	COMDAT @SHA1Input@12
_TEXT	SEGMENT
_len$ = 8
@SHA1Input@12 PROC NEAR					; COMDAT
; _context$ = ecx
; _message_array$ = edx
; Line 37
	mov	eax, DWORD PTR _len$[esp-4]
	push	ebx
	push	esi
	mov	esi, ecx
	mov	ebx, edx
	mov	ecx, DWORD PTR [esi+20]
	lea	edx, DWORD PTR [ecx+eax*8]
; Line 38
	cmp	edx, ecx
	jge	SHORT $L85013
; Line 39
	inc	DWORD PTR [esi+24]
$L85013:
; Line 42
	mov	ecx, eax
	mov	DWORD PTR [esi+20], edx
	sar	ecx, 29					; 0000001dH
	add	DWORD PTR [esi+24], ecx
; Line 44
	cmp	eax, 64					; 00000040H
	jl	SHORT $L85016
	push	ebp
	push	edi
	shr	eax, 6
	lea	edi, DWORD PTR [esi+28]
	mov	ebp, eax
$L85014:
; Line 45
	push	64					; 00000040H
	push	ebx
	push	edi
	call	_memcpy
	add	esp, 12					; 0000000cH
; Line 46
	mov	ecx, esi
	call	@SHA1ProcessMessageBlock@4
; Line 47
	add	ebx, 64					; 00000040H
	dec	ebp
	jne	SHORT $L85014
	pop	edi
	pop	ebp
$L85016:
	pop	esi
	pop	ebx
; Line 49
	ret	4
@SHA1Input@12 ENDP
; Function compile flags: /Ogsy
_TEXT	ENDS
;	COMDAT @SHA1ProcessMessageBlock@4
_TEXT	SEGMENT
_i$ = -4
_W$ = -336
_A$ = -12
_C$ = -16
_D$ = -8
@SHA1ProcessMessageBlock@4 PROC NEAR			; COMDAT
; _context$ = ecx
; Line 52
	push	ebp
	mov	ebp, esp
	sub	esp, 336				; 00000150H
	push	ebx
	push	esi
	push	edi
	mov	eax, ecx
; Line 60
	push	16					; 00000010H
	lea	edi, DWORD PTR _W$[ebp]
	lea	esi, DWORD PTR [eax+28]
	pop	ecx
	rep movsd
; Line 62
	push	64					; 00000040H
	lea	ecx, DWORD PTR _W$[ebp+8]
	pop	edx
$L85033:
; Line 63
	mov	esi, DWORD PTR [ecx+44]
	xor	esi, DWORD PTR [ecx+24]
	xor	esi, DWORD PTR [ecx-8]
	xor	esi, DWORD PTR [ecx]
	add	ecx, 4
	dec	edx
	mov	DWORD PTR [ecx+52], esi
	jne	SHORT $L85033
; Line 66
	mov	ecx, DWORD PTR [eax]
; Line 69
	mov	edx, DWORD PTR [eax+12]
	mov	edi, DWORD PTR [eax+8]
; Line 72
	and	DWORD PTR _i$[ebp], 0
	mov	DWORD PTR _A$[ebp], ecx
	mov	ecx, DWORD PTR [eax+4]
	mov	DWORD PTR _D$[ebp], edx
	mov	edx, DWORD PTR [eax+16]
	mov	DWORD PTR _C$[ebp], edi
$L85036:
; Line 73
	mov	esi, ecx
	mov	ebx, DWORD PTR _A$[ebp]
	not	esi
	and	esi, DWORD PTR _D$[ebp]
	and	edi, ecx
	shl	ebx, 5
	or	esi, edi
	mov	edi, DWORD PTR _A$[ebp]
	sar	edi, 27					; 0000001bH
	or	edi, ebx
	add	esi, edi
	mov	edi, DWORD PTR _i$[ebp]
	add	esi, DWORD PTR _W$[ebp+edi*4]
; Line 75
	mov	edi, DWORD PTR _C$[ebp]
	lea	esi, DWORD PTR [esi+edx+1518500249]
	mov	edx, DWORD PTR _D$[ebp]
	mov	DWORD PTR _D$[ebp], edi
; Line 76
	mov	edi, ecx
	shl	edi, 30					; 0000001eH
	sar	ecx, 2
	or	edi, ecx
	inc	DWORD PTR _i$[ebp]
; Line 77
	mov	ecx, DWORD PTR _A$[ebp]
	cmp	DWORD PTR _i$[ebp], 20			; 00000014H
	mov	DWORD PTR _C$[ebp], edi
; Line 78
	mov	DWORD PTR _A$[ebp], esi
	jl	SHORT $L85036
; Line 81
	mov	DWORD PTR _i$[ebp], 20			; 00000014H
$L85039:
; Line 82
	mov	ebx, esi
	sar	ebx, 27					; 0000001bH
	shl	esi, 5
	or	ebx, esi
	mov	esi, DWORD PTR _D$[ebp]
	xor	esi, edi
	xor	esi, ecx
	add	ebx, esi
	mov	esi, DWORD PTR _i$[ebp]
	add	ebx, DWORD PTR _W$[ebp+esi*4]
	lea	esi, DWORD PTR [ebx+edx+1859775393]
; Line 83
	mov	edx, DWORD PTR _D$[ebp]
; Line 84
	mov	DWORD PTR _D$[ebp], edi
; Line 85
	mov	edi, ecx
	shl	edi, 30					; 0000001eH
	sar	ecx, 2
	or	edi, ecx
	inc	DWORD PTR _i$[ebp]
; Line 86
	mov	ecx, DWORD PTR _A$[ebp]
	cmp	DWORD PTR _i$[ebp], 40			; 00000028H
; Line 87
	mov	DWORD PTR _A$[ebp], esi
	jl	SHORT $L85039
	mov	DWORD PTR _C$[ebp], edi
; Line 90
	mov	DWORD PTR _i$[ebp], 40			; 00000028H
$L85042:
; Line 91
	mov	ebx, edi
	and	edi, ecx
	or	ebx, ecx
	and	ebx, DWORD PTR _D$[ebp]
	or	ebx, edi
	mov	edi, esi
	sar	edi, 27					; 0000001bH
	shl	esi, 5
	or	edi, esi
	mov	esi, DWORD PTR _i$[ebp]
	add	ebx, edi
; Line 93
	mov	edi, DWORD PTR _C$[ebp]
	add	ebx, DWORD PTR _W$[ebp+esi*4]
	lea	esi, DWORD PTR [ebx+edx-1894007588]
	mov	edx, DWORD PTR _D$[ebp]
	mov	DWORD PTR _D$[ebp], edi
; Line 94
	mov	edi, ecx
	shl	edi, 30					; 0000001eH
	sar	ecx, 2
	or	edi, ecx
	inc	DWORD PTR _i$[ebp]
; Line 95
	mov	ecx, DWORD PTR _A$[ebp]
	cmp	DWORD PTR _i$[ebp], 60			; 0000003cH
	mov	DWORD PTR _C$[ebp], edi
; Line 96
	mov	DWORD PTR _A$[ebp], esi
	jl	SHORT $L85042
; Line 99
	mov	DWORD PTR _i$[ebp], 60			; 0000003cH
$L85045:
; Line 100
	mov	ebx, esi
	sar	ebx, 27					; 0000001bH
	shl	esi, 5
	or	ebx, esi
	mov	esi, DWORD PTR _D$[ebp]
	xor	esi, edi
	xor	esi, ecx
	add	ebx, esi
	mov	esi, DWORD PTR _i$[ebp]
	add	ebx, DWORD PTR _W$[ebp+esi*4]
	lea	esi, DWORD PTR [ebx+edx-899497514]
; Line 101
	mov	edx, DWORD PTR _D$[ebp]
; Line 102
	mov	DWORD PTR _D$[ebp], edi
; Line 103
	mov	edi, ecx
	shl	edi, 30					; 0000001eH
	sar	ecx, 2
	or	edi, ecx
	inc	DWORD PTR _i$[ebp]
; Line 104
	mov	ecx, DWORD PTR _A$[ebp]
	cmp	DWORD PTR _i$[ebp], 80			; 00000050H
; Line 105
	mov	DWORD PTR _A$[ebp], esi
	jl	SHORT $L85045
; Line 108
	mov	ebx, DWORD PTR [eax]
	add	ebx, esi
; Line 109
	mov	esi, DWORD PTR [eax+4]
	add	esi, ecx
; Line 110
	mov	ecx, DWORD PTR [eax+8]
	add	ecx, edi
	mov	DWORD PTR [eax+4], esi
; Line 111
	mov	esi, DWORD PTR _D$[ebp]
	mov	DWORD PTR [eax+8], ecx
	mov	ecx, DWORD PTR [eax+12]
	pop	edi
	add	ecx, esi
	mov	DWORD PTR [eax], ebx
	mov	DWORD PTR [eax+12], ecx
; Line 112
	mov	ecx, DWORD PTR [eax+16]
	add	ecx, edx
	pop	esi
	mov	DWORD PTR [eax+16], ecx
	pop	ebx
; Line 113
	leave
	ret	0
@SHA1ProcessMessageBlock@4 ENDP
_TEXT	ENDS
PUBLIC	@SHA1Reset@4
PUBLIC	@SHA1Init@4
; Function compile flags: /Ogsy
;	COMDAT @SHA1Reset@4
_TEXT	SEGMENT
@SHA1Reset@4 PROC NEAR					; COMDAT
; _n$ = ecx
; Line 117
	imul	ecx, 92					; 0000005cH
	add	ecx, OFFSET FLAT:_sgSHA1
	jmp	@SHA1Init@4
@SHA1Reset@4 ENDP
; Function compile flags: /Ogsy
_TEXT	ENDS
;	COMDAT @SHA1Init@4
_TEXT	SEGMENT
@SHA1Init@4 PROC NEAR					; COMDAT
; _context$ = ecx
; Line 122
	and	DWORD PTR [ecx+20], 0
; Line 123
	and	DWORD PTR [ecx+24], 0
; Line 124
	mov	DWORD PTR [ecx], 1732584193		; 67452301H
; Line 125
	mov	DWORD PTR [ecx+4], -271733879		; efcdab89H
; Line 126
	mov	DWORD PTR [ecx+8], -1732584194		; 98badcfeH
; Line 127
	mov	DWORD PTR [ecx+12], 271733878		; 10325476H
; Line 128
	mov	DWORD PTR [ecx+16], -1009589776		; c3d2e1f0H
; Line 129
	ret	0
@SHA1Init@4 ENDP
_TEXT	ENDS
END
